
----------

# Case Study
**ã‚±ãƒ¼ã‚¹ã‚¹ã‚¿ãƒ‡ã‚£**

## Table of Contents
- [unix](#unix)
- [Markdown](#markdown)
- [compress](#compress) zip tar
- [mice8992](#mice8992) Discovering Patterns in the Microbiome
- [my_project](#my_project)
- [NCBI Datasets](#ncbi-datasets)
  - [Downloading data](#Downloading-data)
  - [Inspecting data](#Inspecting-data)
  - [Questions](#questions)
- [Human chromosome 22](#human-chromosome-22) ãƒ’ãƒˆ22ç•ªæŸ“è‰²ä½“
- [Mouse reference genome](#mouse-reference-genome) ãƒã‚¦ã‚¹å‚ç…§ã‚²ãƒãƒ 
- [UniProtKB/Swiss-Prot protein sequence database](#uniprot_sprot) ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹
- [NCBI ASSEMBLY_REPORTS](#ncbi-assembly_reports)
- [](#)

----------
## unix

- 2023 [Bioinformatics Workbook](https://bioinformaticsworkbook.org/)
  - [Introduction to Unix](https://bioinformaticsworkbook.org/Appendix/Unix/unix-basics-1.html)
  - [Unix CheatSheet](https://bioinformaticsworkbook.org/Appendix/Unix/UnixCheatSheet.html)
- 2000-2023 [Learning the shell - Lesson 2: Navigation](https://linuxcommand.org/lc3_lts0020.php)
- 2018 [Introduction to UNIX Commands](https://esslab.jp/~ess/teaching/2018/intro_c_program/unix_command_eng.pdf)
- 2014 [How To Use cd, pwd, and ls to Explore the File System on a Linux Server | DigitalOcean](https://www.digitalocean.com/community/tutorials/how-to-use-cd-pwd-and-ls-to-explore-the-file-system-on-a-linux-server)

2017å¹´11æœˆ22æ—¥
[ã‚¿ãƒ¼ãƒŸãƒŠãƒ« ä½œæ¥­åŠ¹ç‡åŒ– tipsé›†](https://qiita.com/shizuma/items/86470203ac8ea6b4d53f)
ã‚·ãƒ§ãƒ¼ãƒˆã‚«ãƒƒãƒˆç·¨
ã‚³ãƒãƒ³ãƒ‰ç·¨

2017å¹´03æœˆ26æ—¥
[Macã®Terminalã§ã®ã‚·ãƒ§ãƒ¼ãƒˆã‚«ãƒƒãƒˆã‚­ãƒ¼](https://qiita.com/orangeboy/items/a51730ec2d157bf064d6)

ã‚«ãƒ¼ã‚½ãƒ«ç§»å‹•
|å…¥åŠ›|åŠ¹æœ|
|:---------|:---------|
| Ctrl + a	|	è¡Œå…ˆé ­ã«ç§»å‹• |
| Ctrl + e	|	è¡Œæœ«å°¾ã«ç§»å‹• |
| option + â†’	|	ï¼‘å˜èªåˆ†å³ã«ç§»å‹• |
| option + â†	|	ï¼‘å˜èªåˆ†å·¦ã«ç§»å‹• |

Jun 1, 2023
[13 Linux Terminal Shortcuts Every Power Linux User Must Know](https://linuxhandbook.com/linux-shortcuts/)

|||
|:---------|:---------|
| Tab | automatically complete what you were typing, or show all the possible results |
| Ctrl + C | stop (terminate) a running program immediately |
| Ctrl + A | move the cursor to the beginning of the line |
| Ctrl + E | move the cursor to the end of the line |
|||

[ã€Œls -lã€ã‚³ãƒãƒ³ãƒ‰ã®è¡¨ç¤ºã‹ã‚‰ãƒ•ã‚¡ã‚¤ãƒ«ã®å±æ€§ã‚’ç†è§£ã—ã‚ˆã† (1/4)ï¼šâ€œå¿œç”¨åŠ›â€ã‚’ã¤ã‘ã‚‹ãŸã‚ã®Linuxå†å…¥é–€ï¼ˆ9ï¼‰ - ï¼ IT](http://www.atmarkit.co.jp/ait/articles/1605/18/news015.html)

[10 Examples of chmod command in UNIX Linux](https://javarevisited.blogspot.com/2012/03/10-example-of-chmod-command-in-unix.html)

![https://javarevisited.blogspot.com/2012/03/10-example-of-chmod-command-in-unix.html](https://1.bp.blogspot.com/-RzUG1frbLvw/XbVnX6AYBpI/AAAAAAAAbbM/h7HpiDW-F8Emd2C0-dULpC9RzP4n8Dh1ACLcBGAsYHQ/s400/fig_permissions_chmod%2Bcommand.jpg)

----------
## Markdown

- https://en.wikipedia.org/wiki/Markdown
- https://ja.wikipedia.org/wiki/Markdown
- https://daringfireball.net/projects/markdown/syntax
Daring Fireball: Markdown Syntax Documentation
- https://www.markdownguide.org/basic-syntax/
- https://docs.github.com/ja/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax
ã¯ã˜ã‚ã«/GitHub ã§ã®åŸ·ç­†/GitHub ã§æ›¸ãè¾¼ã¿ã‚’é–‹å§‹ã™ã‚‹/åŸºæœ¬çš„ãªæ›¸å¼è¨­å®šã®æ§‹æ–‡
- https://github.com/vsbuffalo/bds-files/tree/master/chapter-02-bioinformatics-projects

Markdown syntax
```

*emphasis*

**bold**

`inline code`

<http://website.com/link>

[link text](http:// website.com/link)
```

----------
## compress

Let's create a project directory (`zmays-snps/`) and create empty data files with sample name (`zmaysA, zmaysB, zmaysC`) and read pair (`R1` and `R2`):  
ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆãƒ»ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒª`zmays-snps/`ã‚’ä½œæˆã—ã€3ã¤ã®ã‚µãƒ³ãƒ—ãƒ«ï¼ˆ`zmaysA, zmaysB, zmaysC`ï¼‰æ¯ã«ãƒšã‚¢ï¼ˆ`R1, R2`ï¼‰ã®ç©ºãƒ‡ãƒ¼ã‚¿ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ä½œæˆã™ã‚‹:  
```
cd ~/projects/
mkdir -p zmays-snps/{data/seqs,scripts,analysis}
touch zmays-snps/data/seqs/zmays{A,B,C}_R{1,2}.fastq
cd zmays-snps/data/
```

Compress the project directory (`zmays-snps/`):  
ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆãƒ»ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒª (`zmays-snps/`) ã‚’åœ§ç¸®ã™ã‚‹:  
```
zip -r zmays-snps.zip zmays-snps/
# or
tar -czvf zmays-snps.tar.gz zmays-snps/
```

Use `man` (e.g. `man zip` and `man tar`) to display the on-line manual pages. press *q* to quit.  
`man ã‚³ãƒãƒ³ãƒ‰å`ï¼ˆä¾‹ãˆã°ã€`man zip`ã‚„`man tar`ï¼‰ã§ã‚³ãƒãƒ³ãƒ‰ã®ãƒãƒ‹ãƒ¥ã‚¢ãƒ«ã‚’å‚ç…§ã™ã‚‹ã€‚çµ‚äº†ã™ã‚‹ã«ã¯ã€*q*ã‚­ãƒ¼ã‚’æŠ¼ã™ã€‚  
```
$man zip

       zip - package and compress (archive) files

       -r
       --recurse-paths
              Travel the directory structure recursively

       -x files
       --exclude files
              Explicitly exclude the specified files
```

```
$man tar

     tar -- manipulate tape archives

     -c      Create a new archive containing the specified items.

     -x      Extract to disk from the archive.

     -z, --gunzip, --gzip
             (c mode only) Compress the resulting archive with gzip(1).

     -v, --verbose

     -f file, --file file
             Read the archive from or write the archive to the specified file.

     --exclude pattern
             Do not process files or directories that match the specified pattern. 

```

- [How to Compress and Extract Files Using the tar Command on Linux](https://www.howtogeek.com/248780/how-to-compress-and-extract-files-using-the-tar-command-on-linux/)
```
# Compress an Entire Directory or a Single File
tar -czvf archive.tar.gz stuff

# Extract an Archive
tar -xzvf archive.tar.gz
```

----------
## mice8992

Discovering Patterns in the Microbiome

The Github repo for this code is https://github.com/danknights/mice8992-2016

[Microbiome Discovery 4.5: (Optional) UNIX Command Line - YouTube](https://www.youtube.com/watch?v=u2IQQUMeWy8&list=PLOPiWVjg6aTzsA53N19YqJQeZpSCH9QPc&index=6)
11:10

Code in the video (and its modification):
```
# 1:35
cd ~
ls
mkdir mice8992
cd mice8992
# 2:30
git clone https://github.com/danknights/mice8992-2016.git repo
cd repo
clear
ls
# 4:20
cd data/
ls
cd globalgut/
ls
# The files in the directory are different from the video.
unzip seqs.fna.zip
head seqs.fna
head -n 2 seqs.fna
head map.txt
# 6:10
head map.txt | cut -f1-3
# 7:00
cd ..
cd ..
ls
git pull
git status
# 7:45
ls
emacs tmp.txt
# hello world
# to save the file, C-x C-s
# to exit from emacs, C-x C-c
mv tmp.txt ..
mv ../tmp.txt .
rm tmp.txt 
# 8:55
cd data/globalgut/
clear
wc seqs.fna
wc -l seqs.fna
grep ">" seqs.fna
grep ">" seqs.fna | wc -l
grep -c ">" seqs.fna
```

----------
### my_project

ã‚¿ãƒ¼ãƒŸãƒŠãƒ«ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã‚’é–‹ãã€‚  
Open a terminal window.  
```
bash
```

æ—¥ä»˜ã®ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’ä½œæˆã—ã¦ã€ãã®ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã«ç§»å‹•ã™ã‚‹:  
Create a dated directory and change into it:  
```
mkdir $(date +%F); cd $(date +%F)
```

ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’å–å¾—ã™ã‚‹:  
Retrieve a project directory:  
```
wget https://github.com/haruosuz/introBI/raw/main/my_project.zip
unzip my_project.zip
cd my_project
ls
```

`any_template/` ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’ã‚³ãƒ”ãƒ¼:  
Copy the `any_template/` directory:  
```
cp -r any_template a_new_project
```

`a_new_project/` ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã‚’å®Ÿè¡Œ:  
Execute the `a_new_project/` project:  
```
cd a_new_project/
#cat README.md
bash scripts/run.sh
```

`gff_ecoli/` ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã‚’å®Ÿè¡Œ:  
Execute the `gff_ecoli/` project:  
```
cd ..
cd gff_ecoli/
#cat README.md
(bash scripts/run.sh &) >& log.$(date +%F).txt
tail -f log.$(date +%F).txt
# Use `tail -f` to constantly monitor files (use Control-C to stop)
# `tail -f`ã§ãƒ•ã‚¡ã‚¤ãƒ«å‡ºåŠ›ã‚’ç›£è¦–ã™ã‚‹ï¼ˆControl-Cã§å‹•ä½œä¸­ã®ãƒ—ãƒ­ã‚»ã‚¹ã‚’åœæ­¢ï¼‰
```

----------
## NCBI Datasets
A one-stop shop for finding, browsing, and downloading genomic data

Retrieving genome sequence data from the NCBI website.  
NCBIã®ã‚¦ã‚§ãƒ–ã‚µã‚¤ãƒˆã‹ã‚‰ã‚²ãƒãƒ é…åˆ—ãƒ‡ãƒ¼ã‚¿ã‚’å–å¾—ã™ã‚‹ã€‚  

- https://www.ncbi.nlm.nih.gov/datasets/  
Click on the **Genome** tab.  
**Genome** ã‚¿ãƒ–ã‚’ã‚¯ãƒªãƒƒã‚¯ã™ã‚‹ã€‚  
- https://www.ncbi.nlm.nih.gov/datasets/genome/  
you will see a search box [ Selected taxa ], and you can type "Enter one or more taxonomic names" of the sequence that you are looking for in this search box, and then press Enter to search for it. For example, if you want to find the sequence for [*Escherichia coli*](https://en.wikipedia.org/wiki/Escherichia_coli), you would type just organism name "**Escherichia coli**" in the search box and press Enter.  
æ¤œç´¢ãƒœãƒƒã‚¯ã‚¹ [ Selected taxa ] ã«ç”Ÿç‰©åã‚’å…¥åŠ›ã—ã¦Enterã‚­ãƒ¼ã‚’æŠ¼ã™ã€‚ä¾‹ãˆã°ã€[å¤§è…¸èŒ](https://ja.wikipedia.org/wiki/å¤§è…¸èŒ) ã®å­¦å "**Escherichia coli**" ã‚’æ¤œç´¢ã™ã‚‹ã€‚  
- https://www.ncbi.nlm.nih.gov/datasets/genome/?taxon=562  
under **Assembly**, you will see genome sequences; of which, click reference genome (e.g., "ASM886v2") indicated by check mark.  
**Assembly** ã®ä¸‹ã«ã€ã‚²ãƒãƒ é…åˆ—ãƒ‡ãƒ¼ã‚¿ãŒè¡¨ç¤ºã•ã‚Œã‚‹ã€‚ã“ã®ã†ã¡ã€ãƒã‚§ãƒƒã‚¯ãƒãƒ¼ã‚¯ãŒä»˜ã„ãŸå‚ç…§ã‚²ãƒãƒ  (ä¾‹ãˆã°ã€"ASM886v2") ã‚’ã‚¯ãƒªãƒƒã‚¯ã™ã‚‹ã€‚  
- https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000008865.2/  

|          |          | Actions |
|:---------|:---------|:---------|
| NCBI RefSeq assembly |	GCF_000008865.2 |	 |
| Submitted GenBank assembly  |GCA_000008865.2 |	 |

[What are NCBI Assembly accession numbers?](https://support.nlm.nih.gov/knowledgebase/article/KA-03451/en-us)

### RefSeq

- https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000008865.2/  
Click on "See more files on FTP" from the menu on the right (under **Actions**) of the **NCBI RefSeq assembly** "GCF_000008865.2".  
**NCBI RefSeq assembly** "GCF_000008865.2" ã®å³å´ã®ãƒ¡ãƒ‹ãƒ¥ãƒ¼ï¼ˆ**Actions**ã®ä¸‹ï¼‰ã‹ã‚‰"See more files on FTP"ã‚’é¸æŠã—ã¦é–‹ãã€‚  
- https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/008/865/GCF_000008865.2_ASM886v2/
```
GCF_000008865.2_ASM886v2_genomic.gff.gz          2021-08-21 03:36  342K  
md5checksums.txt                                 2024-03-31 07:25  1.1K  
```

### GenBank

- https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000008865.2/  
Click on "See more files on FTP" from the menu on the right (under **Actions**) of the **Submitted GenBank assembly** "GCA_000008865.2".  
**Submitted GenBank assembly** "GCA_000008865.2" ã®å³å´ã®ãƒ¡ãƒ‹ãƒ¥ãƒ¼ï¼ˆ**Actions**ã®ä¸‹ï¼‰ã‹ã‚‰"See more files on FTP"ã‚’é¸æŠã—ã¦é–‹ãã€‚  
- https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/008/865/GCA_000008865.2_ASM886v2/  
```
GCA_000008865.2_ASM886v2_genomic.gff.gz          2021-08-23 16:06  305K  
md5checksums.txt                                 2024-03-31 07:25  1.1K  
```

Right click on the link to the file (*\*_genomic.gff.gz*, *md5checksums.txt*), and select "Copy Link Address".  
ãƒ•ã‚¡ã‚¤ãƒ«ï¼ˆ*\*_genomic.gff.gz*, *md5checksums.txt*ï¼‰ã¸ã®ãƒªãƒ³ã‚¯ã‚’å³ã‚¯ãƒªãƒƒã‚¯ã—ã€ã€Œãƒªãƒ³ã‚¯ã®URLã‚’ã‚³ãƒ”ãƒ¼ (Copy Link)ã€ã™ã‚‹ã€‚  

ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªå†…ã®ãƒ•ã‚¡ã‚¤ãƒ«
[What is the file content within each specific assembled genome directory?](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/policies-annotation/genomeftp/#what-is-the-file-content-within-each-specific-assembled-genome-directory)  
https://ftp.ncbi.nlm.nih.gov/genomes/all/README.txt
```
File formats and content:

   *_genomic.gff.gz file
       Annotation of the genomic sequence(s) in Generic Feature Format Version 3
       (GFF3). 

   md5checksums.txt file
       file checksums are provided for all data files in the directory
```

### Downloading data
ãƒ‡ãƒ¼ã‚¿ã®ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰

[ã‚¿ãƒ¼ãƒŸãƒŠãƒ«](https://techacademy.jp/magazine/5155)ã‚’é–‹ãã€`bash`ã‚’èµ·å‹•ã—ã€æ—¥ä»˜ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’ä½œæˆã—ã¦ç§»å‹•ã™ã‚‹:  
```
# Open a terminal window
bash

# Create a dated directory and change into it
mkdir $(date +%F); cd $(date +%F)
```

`wget`ã‚³ãƒãƒ³ãƒ‰ã‚’ä½¿ç”¨ã—ã¦ã€ã‚²ãƒãƒ ã‚¢ãƒãƒ†ãƒ¼ã‚·ãƒ§ãƒ³ã®[GFF](https://github.com/haruosuz/bioinfo/blob/master/README.md#gff)å½¢å¼ãƒ•ã‚¡ã‚¤ãƒ«ï¼ˆ*\*_genomic.gff.gz*ï¼‰ã€[MD5](https://ja.wikipedia.org/wiki/MD5)[ãƒã‚§ãƒƒã‚¯ã‚µãƒ ](https://ja.wikipedia.org/wiki/ãƒã‚§ãƒƒã‚¯ã‚µãƒ )ãƒ•ã‚¡ã‚¤ãƒ«ï¼ˆ*md5checksums.txt*ï¼‰ã‚’ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã™ã‚‹:  
```
# Using `wget` to download the files (*_genomic.gff.gz, md5checksums.txt):
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/008/865/GCA_000008865.2_ASM886v2/GCA_000008865.2_ASM886v2_genomic.gff.gz
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/008/865/GCA_000008865.2_ASM886v2/md5checksums.txt
```

ãƒã‚§ãƒƒã‚¯ã‚µãƒ ã‚’ç¢ºèªã™ã‚‹:  
```
# compare our checksum values with those in "md5checksums.txt" using the md5 program:
grep "_genomic.gff.gz" *md5checksums.txt

md5sum *.gz

# macOS
md5 *.gz
```

*README.md*ãƒ•ã‚¡ã‚¤ãƒ«ã®ä¾‹ã¯æ¬¡ã®é€šã‚Š:  
The *README.md* file would look like:  
```
## Genome and Annotation Data

Escherichia coli O157:H7 str. Sakai, Complete Genome (GenBank: GCA_000008865.2) data were downloaded on 2025-10-28 using:

    wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/008/865/GCA_000008865.2_ASM886v2/GCA_000008865.2_ASM886v2_genomic.gff.gz

## MD5 checksum
- MD5 (GCA_000008865.2_ASM886v2_genomic.gff.gz) = 447259f763817b5d58704966d72b377e
```

`gunzip`ã‚³ãƒãƒ³ãƒ‰ã§ãƒ•ã‚¡ã‚¤ãƒ«ã‚’å±•é–‹ã™ã‚‹:  
```
# Decompress files with the command `gunzip`:
gunzip *.gz
```

### Inspecting data
ãƒ‡ãƒ¼ã‚¿ã®æ¤œæŸ»

Using core Unix tools (`head, tail, wc, ls, awk, cut, grep, sort, uniq`) to analyze Annotation of the genomic sequence (*\*_genomic.gff* file).  
Unixã‚³ãƒãƒ³ãƒ‰ï¼ˆ`head, tail, wc, ls, awk, cut, grep, sort, uniq`ï¼‰ã‚’ç”¨ã„ã¦ã€ã‚²ãƒãƒ ã®ã‚¢ãƒãƒ†ãƒ¼ã‚·ãƒ§ãƒ³ï¼ˆ*\*_genomic.gff*ãƒ•ã‚¡ã‚¤ãƒ«ï¼‰ã‚’è§£æã™ã‚‹ã€‚  

```
# p.134 | Plain-Text Data Summary Information with wc, ls, and awk
# è¨³æ›¸147é  | 7.3.3ã€€wcã€lsã€awkã«ã‚ˆã‚‹ãƒ—ãƒ¬ãƒ¼ãƒ³ãƒ†ã‚­ã‚¹ãƒˆãƒ‡ãƒ¼ã‚¿ã®è¦ç´„æƒ…å ±

# `ls -lh`ã§ãƒ•ã‚¡ã‚¤ãƒ«ã®ã‚µã‚¤ã‚ºã‚’ç¢ºèªã™ã‚‹:  
# `ls -lh` reports human-readable file sizes
ls -lh

# `wc -l`ã§ç·è¡Œæ•°ã‚’ã‚«ã‚¦ãƒ³ãƒˆã™ã‚‹:  
# `wc -l` outputs the number of lines
wc -l *

# p.398 | Variables and Command Arguments
# è¨³æ›¸425é  | 12.1.2ã€€å¤‰æ•°ã¨ã‚³ãƒãƒ³ãƒ‰å¼•æ•°

# å¤‰æ•°ã«å€¤ã‚’å‰²ã‚Šå½“ã¦ã‚‹ï¼ˆ`=`ã®å‰å¾Œã«ã‚¹ãƒšãƒ¼ã‚¹ã‚’å…¥ã‚Œãªã„ï¼‰:  
# create a variable and assign it a value with (do not use spaces around the equals sign!):  
GFF=GCA_000008865.2_ASM886v2_genomic.gff

# å¤‰æ•°ã®å€¤ã«ã‚¢ã‚¯ã‚»ã‚¹ã™ã‚‹ã«ã¯ã€å¤‰æ•°åã®å‰ã«ãƒ‰ãƒ«è¨˜å·ã‚’ä»˜ã‘ã‚‹:  
# To access a variableâ€™s value, we use a dollar sign in front of the variableâ€™s name:  
echo "${GFF}"

# p.129 | Inspecting Data with Head and Tail
# è¨³æ›¸142é  | 7.3.1 head ã¨ tail ã«ã‚ˆã‚‹ãƒ‡ãƒ¼ã‚¿ã®æ¤œæŸ»

# `head`ã§å…ˆé ­éƒ¨åˆ†ã‚’è¡¨ç¤ºã™ã‚‹:  
# look at the top of a file
head -n 10 "${GFF}"

# `tail`ã§æœ«å°¾éƒ¨åˆ†ã‚’è¡¨ç¤ºã™ã‚‹:  
# look at the end of a file
tail -n 3 "${GFF}"

# p.47 | Pipes in Action: Creating Simple Programs with Grep and Pipes
# è¨³æ›¸53é  | 3.3.1ã€€ãƒ‘ã‚¤ãƒ—ã®å‹•ä½œï¼šgrepã¨ãƒ‘ã‚¤ãƒ—ã«ã‚ˆã‚‹ç°¡å˜ãªãƒ—ãƒ­ã‚°ãƒ©ãƒ ã®ä½œæˆ

# p.140 | The All-Powerful Grep
# è¨³æ›¸153é  | 7.3.6ã€€å¼·åŠ›ãªãƒ„ãƒ¼ãƒ«grep

# `grep`ã§ãƒ‘ã‚¿ãƒ¼ãƒ³ã«ãƒãƒƒãƒã—ãŸè¡Œã‚’æŠ½å‡ºã™ã‚‹:  
# use grep to extract lines matching the pattern:  
## GFF header lines begin with the "#" character.
grep "^#" "${GFF}"

# ãƒ‘ã‚¤ãƒ—ã§ãƒ—ãƒ­ã‚°ãƒ©ãƒ ã®å…¥å‡ºåŠ›ã‚’ã¤ãªãã€‚
# Pipe the standard output to the next command with the pipe character (|):  
grep "^#" "${GFF}" | wc -l

# `grep -c`ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã§ã€ãƒ‘ã‚¿ãƒ¼ãƒ³ã«ãƒãƒƒãƒã—ãŸè¡Œæ•°ã‚’è¡¨ç¤ºã™ã‚‹:  
# use grep to count (the -c option stands for count) the number of lines matching the pattern:  
grep -c "^#" "${GFF}"

# `grep -v`ã§"#"ã§å§‹ã¾ã‚‹è¡Œã‚’å‰Šé™¤ã™ã‚‹:  
# exclude lines that begin with "#":  
grep -v "^#" "${GFF}" | head -n 3

# p.138 | Working with Column Data with cut and Columns
# è¨³æ›¸151é  | 7.3.4 cut ã«ã‚ˆã‚‹åˆ—ãƒ‡ãƒ¼ã‚¿ã®æ“ä½œ

# `grep -v`ã§"#"ã§å§‹ã¾ã‚‹è¡Œã‚’å‰Šé™¤ã—ã€`cut`ã§1,4,5åˆ—ï¼ˆé…åˆ—ã®åå‰ã€é–‹å§‹ä½ç½®ã€çµ‚äº†ä½ç½®ï¼‰ã‚’æŠ½å‡ºã™ã‚‹:  
# chop off the metadata rows using `grep`, and then use `cut` to extract
# the first, fourth, and fifth columns (chromosome, start, end):
grep -v "^#" "${GFF}" | cut -f1,4,5 | head -n 3

# p.147 | Sorting Plain-Text Data with Sort
# è¨³æ›¸160é  | 7.3.8ã€€sortã«ã‚ˆã‚‹ãƒ—ãƒ¬ãƒ¼ãƒ³ãƒ†ã‚­ã‚¹ãƒˆãƒ‡ãƒ¼ã‚¿ã®ä¸¦ã¹æ›¿ãˆ

# p.152 | Finding Unique Values in Uniq
# è¨³æ›¸165é  | 7.3.9ã€€uniqã‚³ãƒãƒ³ãƒ‰ã§ä¸€æ„ã®å€¤ã‚’è¦‹ã¤ã‘ã‚‹

# Unixã‚³ãƒãƒ³ãƒ‰ï¼ˆ`grep, cut, sort, uniq`ï¼‰ã‚’çµ„ã¿åˆã‚ã›ã¦ã€è¡¨å½¢å¼ãƒ‡ãƒ¼ã‚¿ã®åˆ—ã‚’è¦ç´„ã™ã‚‹:  
# combine Unix tools (`grep, cut, sort, uniq`) to summarize columns of tabular data:
grep -v "^#" "${GFF}" | cut -f3 | sort | uniq -c

# +é–ã¨-é–(7åˆ—ç›®)ã®ç‰¹å¾´é ˜åŸŸ(3åˆ—ç›®)ã‚’ã‚«ã‚¦ãƒ³ãƒˆã™ã‚‹:  
# create lines from multiple columns to count combinations, like
# how many of each feature (column 3) are on each strand (column 7):
grep -v "^#" "${GFF}" | cut -f3,7 | sort | uniq -c

# p.157 | Text Processing with Awk
# è¨³æ›¸170é  | 7.3.11ã€€AWKã«ã‚ˆã‚‹ãƒ†ã‚­ã‚¹ãƒˆå‡¦ç†

# `grep -o`ã§ãƒ‘ã‚¿ãƒ¼ãƒ³ã®ä¸€è‡´ã™ã‚‹éƒ¨åˆ†ã ã‘ã‚’æŠ½å‡ºã™ã‚‹ã€‚  
# `grep -o` extract only the matching part of the pattern.  

# rRNAéºä¼å­ï¼ˆ16Sã€23Sã€5Sï¼‰ã®æ•°ã‚’ã‚«ã‚¦ãƒ³ãƒˆã™ã‚‹:  
# Count rRNA genes (16S, 23S, 5S):  
awk -F"\t" '$3 ~ /rRNA/ { print $0 }' "${GFF}" | grep -E -o 'product=.+' | sort | uniq -c

# rRNAéºä¼å­ã®é–‹å§‹ä½ç½®ã¨çµ‚äº†ä½ç½®ã‚’æŠ½å‡ºã™ã‚‹:  
# Extract start positions and end positions of rRNA genes (16S, 23S, 5S):  
awk -F"\t" '$3 ~ /rRNA/ { print $0 }' "${GFF}" | cut -f4,5,9 | cut -d"=" -f1,8

# ã‚¿ãƒ³ãƒ‘ã‚¯è³ªã‚³ãƒ¼ãƒ‰é…åˆ—ï¼ˆCDSï¼‰ã«ã¤ã„ã¦ã€é•·ã•ï¼ˆçµ‚äº†ä½ç½® - é–‹å§‹ä½ç½®ï¼‰ã®åˆ—ã‚’è¿½åŠ ã—ã€æ•°å€¤é †ã«ã‚½ãƒ¼ãƒˆã—ã€æœ«å°¾ã‚’è¦‹ã‚‹:  
# add a column with the feature length (end position - start position) for only "CDS"
awk -F"\t" '$3 ~ /CDS/ { print $5 - $4 "\t" $0 }' "${GFF}" | sort -k1,1n | tail -n 1
```

å‡ºåŠ›ä¾‹:  
```
$grep -v "^#" "${GFF}" | cut -f3 | sort | uniq -c
5290 CDS
 126 exon
5243 gene
 136 pseudogene
  22 rRNA
   3 region
  65 repeat_region
  24 sequence_feature
 103 tRNA
   1 tmRNA

$awk -F"\t" '$3 ~ /rRNA/ { print $0 }' "${GFF}" | grep -E -o 'product=.+' | sort | uniq -c
   7 product=16S ribosomal RNA
   7 product=23S ribosomal RNA
   8 product=5S ribosomal RNA
```

### Questions
- How many genes ([CDS](https://en.wikipedia.org/wiki/Coding_region), [rRNA](https://en.wikipedia.org/wiki/Ribosomal_RNA), [tRNA](https://en.wikipedia.org/wiki/Transfer_RNA)) are in the genome?  
ã‚²ãƒãƒ ã«å«ã¾ã‚Œã‚‹éºä¼å­ï¼ˆgeneï¼‰ã®æ•°ã¯ï¼Ÿã‚¿ãƒ³ãƒ‘ã‚¯è³ªã‚’ã‚³ãƒ¼ãƒ‰ã™ã‚‹é…åˆ—ï¼ˆ[CDS](https://ja.wikipedia.org/wiki/ã‚³ãƒ¼ãƒ‡ã‚£ãƒ³ã‚°é ˜åŸŸ)ï¼‰ã€[ribosomal RNA (rRNA)](https://ja.wikipedia.org/wiki/ãƒªãƒœã‚½ãƒ¼ãƒ RNA)ã€[transfer RNA (tRNA)](https://ja.wikipedia.org/wiki/è»¢ç§»RNA) éºä¼å­ã®æ•°ã¯ï¼Ÿ
- In the genome, rRNA genes are arranged in the order 16S-23S-5S? The distance between each gene is â‰¤1500â€‰bp?  
ãƒªãƒœã‚½ãƒ¼ãƒ RNAã¯ã€åŸæ ¸ç”Ÿç‰©ã§ã¯ã€16Sã€23Sã€5Sã®é †ã«ä¸¦ã‚“ã ã‚ªãƒšãƒ­ãƒ³æ§‹é€ ã‚’æŒã£ã¦ã„ã‚‹ï¼Ÿ
16S:23S:5S = 1:1:1ã§å­˜åœ¨ï¼Ÿ
16Sã¨23Sã®è·é›¢ã¯1500å¡©åŸºä»¥ä¸‹ï¼Ÿ
- What is the length in nucleotides of the gene sequence? What is the maximum length (bp) of CDS?  
éºä¼å­ã®é•·ã•ã¯ï¼Ÿã‚¿ãƒ³ãƒ‘ã‚¯è³ªã‚’ã‚³ãƒ¼ãƒ‰ã™ã‚‹é…åˆ—ï¼ˆCDSï¼‰ã®é•·ã•ã®æœ€å¤§å€¤ã¯ï¼Ÿ  

### References
- DDBJ å…¬é–‹å½¢å¼ Flat file [Feature key](https://www.ddbj.nig.ac.jp/ddbj/features.html) | DDBJ flat file format [Feature key](https://www.ddbj.nig.ac.jp/ddbj/features-e.html)
CDS
rRNA
tRNA
- [Kumar et al. Sci Rep. 2019 Jan 17 "The mitochondrial genome of the marigold thrips, N. samayunkur."](https://pmc.ncbi.nlm.nih.gov/articles/PMC6336932/figure/Fig1/)
  - [Neohydatothrips samayunkur mitochondrion, complete genome](https://www.ncbi.nlm.nih.gov/nuccore/MF991901)
 ORIGIN 
 // 
- [Hu et al. PLoS Genet. 2021 Nov 29 "An optimal growth law for RNA composition and its partial implementation through ribosomal and tRNA gene locations in bacterial genomes"](https://pubmed.ncbi.nlm.nih.gov/34843465/)
  - As shown in Fig 3A and 3B (orange points), we found that in fast-growing species, rRNA and tRNA genes are generally located in the vicinity of oriC, ... Moreover, we found that both rRNA and tRNA genes tend to be located ever closer to oriC with increasing Î¼max (...
  - [Fig 3](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8659690/figure/pgen.1009939.g003/)
- [Huang et al. Nat Commun. 2020 Jun 9 "Ribosome engineering reveals the importance of 5S rRNA autonomy for ribosome assembly"](https://pubmed.ncbi.nlm.nih.gov/32518240/)
- [Brewer et al. ISME J. 2020 Feb "Unlinked rRNA genes are widespread among bacteria and archaea."](https://www.nature.com/articles/s41396-019-0552-3)
  - The frequency of unlinked rRNA genes may reflect meaningful life history traits, as they tend to be associated with a mix of slow-growing free-living species and intracellular species.
  - [Fig. 1](https://www.nature.com/articles/s41396-019-0552-3/figures/1) In most bacteria and archaea, rRNA genes are arranged in the order 16S-23S-5S, and are transcribed and regulated as a single unit. However, in some cases, the 16S is separated from the 23S and 5S, and is referred to as â€œunlinkedâ€
  - rRNA pairs were classified as â€œunlinkedâ€ if the distance between each gene was >1500â€‰bp, â€œlinkedâ€ if the distance was â‰¤1500â€‰bp.  
å¤§éƒ¨åˆ†ã®åŸæ ¸ç”Ÿç‰©ã§ã¯rRNAéºä¼å­ã¯16S-23S-5Sã®é †åºã§é…ç½®ã•ã‚Œã€ã¾ã¨ã‚ã¦è»¢å†™ã•ã‚Œã‚‹ãŒã€ä¸€éƒ¨ã®ã‚±ãƒ¼ã‚¹ã§ã¯ã€16SãŒ23S-5Sã‹ã‚‰ï¼ˆ>1500å¡©åŸºã®è·é›¢ã§ï¼‰é›¢ã‚Œã¦ã„ã‚‹ã€‚ã“ã® â€œunlinkedâ€ rRNAéºä¼å­ã¯ã€åŸæ ¸ç”Ÿç‰©ã«åºƒãè¦‹ã‚‰ã‚Œã€ãã®é »åº¦ã¯ç”Ÿæ´»å²ï¼ˆé…ã„å¢—æ®–ã¨ç´°èƒå†…å…±ç”Ÿï¼‰ã‚’åæ˜ ã™ã‚‹ã€‚
- [Vieira-Silva & Rocha (2010) PLoS Genet "The Systemic Imprint of Growth and Its Uses in Ecological (Meta)Genomics"](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2797632/)
  - [Figure 1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2797632/figure/pgen-1000808-g001/) Genomic signatures correlated to minimum generation time (d) for 214 prokaryotes.
Correlation between d and (A) the number of rRNA operons in the genome, (B) the relative distance from the origin of replication to rRNA genes (excluding species with no retrievable origin), 0.5 corresponds to half the replicon, (C,D) codon usage bias indices Î”ENCâ€² [35] and S [46].  
rRNAã‚ªãƒšãƒ­ãƒ³æ•°ãŒå¤šã„ã»ã©ã€rRNAéºä¼å­ã®è¤‡è£½èµ·ç‚¹ã‹ã‚‰ã®è·é›¢ãŒå°ã•ã„ã»ã©ã€ã‚³ãƒ‰ãƒ³ä½¿ç”¨ãƒã‚¤ã‚¢ã‚¹ãŒå¼·ã„ã»ã©ã€åŸæ ¸ç”Ÿç‰©ã®æœ€å°å€åŠ æ™‚é–“ãŒå°ã•ã„ã€‚  
- [Sharp et al. (2005) Nucleic Acids Res "Variation in the strength of selected codon usage bias among bacteria"](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC549432/)   
  - species exposed to selection for rapid growth have more rRNA operons, more tRNA genes and more strongly selected codon usage bias.  
rRNAã‚ªãƒšãƒ­ãƒ³æ•°/tRNAéºä¼å­ã‚³ãƒ”ãƒ¼æ•°ãŒå¤šã„ã»ã©ã€ã‚³ãƒ‰ãƒ³ä½¿ç”¨ãƒã‚¤ã‚¢ã‚¹ãŒå¼·ã„ã»ã©ã€ç´°èŒã®å¢—æ®–é€Ÿåº¦ãŒé€Ÿã„ã€‚ 
  - [Table 1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC549432/table/tbl1/)
The 80 bacterial genome sequences analysed | Gene numbers: rRNA, tRNA, ORF  

----------
## Human chromosome 22
**ãƒ’ãƒˆ22ç•ªæŸ“è‰²ä½“**

- 2020/09/02 | UCSC Genome Browser | TogoTV https://www.youtube.com/playlist?list=PL18C454AF92F9B012
- 2011å¹´11æœˆ28æ—¥ [ã‚·ãƒ§ãƒ¼ãƒˆãƒªãƒ¼ãƒ‰ã®æ†‚é¬± - æ¬¡ä¸–ä»£ã‚·ãƒ¼ã‚±ãƒ³ã‚µãƒ¼: GRCh37ã¨Hg19](http://shortreadbrothers.blogspot.com/2011/11/grch37hg19.html)
- Surfer's-wiki [GRCh37ã¨Hg19ã®é•ã„(å«ã‚€ãƒŸãƒˆã‚³ãƒ³ãƒ‰ãƒªã‚¢)](https://cell-innovation.nig.ac.jp/SurfWiki/GRCh37_Hg19_defference.html)

Retrieving URLs from the UCSC Genome Browser.  
UCSC Genome Browser ã‹ã‚‰URLã‚’å–å¾—ã™ã‚‹ã€‚ 

<http://genome.ucsc.edu>  

- Open [Downloads â†’ Genome Data](https://hgdownload.soe.ucsc.edu/downloads.html).  
[Downloads â†’ Genome Data](http://hgdownload.soe.ucsc.edu/downloads.html) ã‚’é–‹ãã€‚  
- Click on **Human genomes**, and open the **Sequence data by chromosome** section below **Feb. 2009 (GRCh37/hg19)**.  
**Human genomes** ã‚’ã‚¯ãƒªãƒƒã‚¯ã—ã€ **Feb. 2009 (GRCh37/hg19)** ã® **Sequence data by chromosome** ã‚’é–‹ãã€‚
- https://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/

      Name                        Last modified      Size  Description

      chr22.fa.gz                 2009-03-20 09:02   11M  

      md5sum.txt                  2009-03-20 09:49  4.9K  

- Right click on the link to the files, and select "Copy Link Address".  
ãƒ•ã‚¡ã‚¤ãƒ«ã¸ã®ãƒªãƒ³ã‚¯ã‚’å³ã‚¯ãƒªãƒƒã‚¯ã—ã€ã€Œãƒªãƒ³ã‚¯ã®URLã‚’ã‚³ãƒ”ãƒ¼ (Copy Link)ã€ã™ã‚‹ã€‚  

### Downloading data
**ãƒ‡ãƒ¼ã‚¿ã®ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰**

ã‚¿ãƒ¼ãƒŸãƒŠãƒ«ã‚’é–‹ãã€‚  
Open a terminal window.  
```
bash

# ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’ä½œæˆã™ã‚‹:
# make directory:
mkdir -p ~/projects/data/ucsc

# ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’ç§»å‹•ã™ã‚‹:
# change directory:
cd ~/projects/data/ucsc/

# åœ§ç¸®ãƒ•ã‚¡ã‚¤ãƒ« chr22.fa.gz ã‚’ `wget` ã‚’ä½¿ç”¨ã—ã¦ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã™ã‚‹:  
# Download the compressed file chr22.fa.gz using `wget`:  
wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/chr22.fa.gz

# `wget` ã‚’ä½¿ç”¨ã—ã¦ md5sum.txt ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã™ã‚‹:  
# Download the md5sum.txt file using `wget`:  
wget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/md5sum.txt

# MD5 ãƒã‚§ãƒƒã‚¯ã‚µãƒ ã‚’è¨ˆç®—ã™ã‚‹:  
# Calculate the MD5 checksum:  
md5 chr22.fa.gz

# md5sum.txt ãƒ•ã‚¡ã‚¤ãƒ«å†…ã§ "chr22.fa.gz" ã‚’æ¤œç´¢ã™ã‚‹:   
# Search for "chr22.fa.gz" in md5sum.txt:  
grep "chr22.fa.gz" md5sum.txt
```

Markdownãƒ•ã‚¡ã‚¤ãƒ« *README.md* ã®ä¾‹:  
The Markdown file *README.md* file would look like:  
```
## Genome Data

the Feb. 2009 assembly of the human genome (hg19,
GRCh37 Genome Reference Consortium Human Reference 37 (GCA_000001405.1))
was downloaded on Mon Apr 29 11:53:23 JST 2024, using:

    wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/chr22.fa.gz
    wget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/md5sum.txt

## MD5 checksum

    $md5 chr22.fa.gz
    MD5 (chr22.fa.gz) = 2baab477ebcc5a0f4163608ec93e4744

    $grep "chr22.fa.gz" md5sum.txt
    2baab477ebcc5a0f4163608ec93e4744  chr22.fa.gz
```

2:28 AM Â· Oct 25, 2024
https://x.com/brent_p/status/1849503122114347051
brent pedersen on X: "don't use hg38.fa as-is. checkout the references ğŸ˜œ here: https://t.co/UBhkn6zjDx rendered the ipynb (not mine) here: https://t.co/Zv6J1d6XAy in short, use: GRCh38_GIABv3_no_alt_analysis_set_maskedGRC_decoys_MAP2K3_KMT2C_KCNJ18.fasta.gz other updates on the best hg38 reference?" / X

----------
## Mouse reference genome
**ãƒã‚¦ã‚¹å‚ç…§ã‚²ãƒãƒ **

- https://github.com/vsbuffalo/bds-files/tree/master/chapter-06-bioinformatics-data
- 2017.06.07 | 04:20 | Ensemblã®ä½¿ã„æ–¹ ã€œé…åˆ—ã‚’å–å¾—ã™ã‚‹ã€œ2017 | TogoTV https://doi.org/10.7875/togotv.2017.046

[Genome Reference Consortium](https://www.ncbi.nlm.nih.gov/grc)  
Retrieving the mitochondrial (MT) sequence of the mouse reference genome.  
ãƒã‚¦ã‚¹å‚ç…§ã‚²ãƒãƒ ã®ãƒŸãƒˆã‚³ãƒ³ãƒ‰ãƒªã‚¢ (MT) é…åˆ—ã‚’å–å¾—ã™ã‚‹ã€‚  

Retrieving URLs from the Ensembl Genome Browser.  
Ensembl Genome Browser ã‹ã‚‰URLã‚’å–å¾—ã™ã‚‹ã€‚ 

Open in a browser (Firefox or Chrome) and navigate to the following pages by clicking the hyperlinks.  
ãƒ–ãƒ©ã‚¦ã‚¶ï¼ˆFirefox ã¾ãŸã¯ Chromeï¼‰ã§é–‹ãã€ãƒã‚¤ãƒ‘ãƒ¼ãƒªãƒ³ã‚¯ã‚’ã‚¯ãƒªãƒƒã‚¯ã—ã¦ä»¥ä¸‹ã®ãƒšãƒ¼ã‚¸ã«ç§»å‹•ã™ã‚‹ã€‚  

- <https://asia.ensembl.org/index.html>
  - [Mouse](http://asia.ensembl.org/Mus_musculus/Info/Index)
    - [Download FASTA files for genes, cDNAs, ncRNA, proteins](https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/)
      - [dna/](https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/)
- https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/
```
Index of /pub/release-111/fasta/mus_musculus/dna

Name	Last modified	Size	Description

CHECKSUMS	2023-10-19 17:06	4.2K	 

Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz	2023-10-04 12:37	5.2K	 

README	2023-10-04 12:39	4.9K	 
```
Right click on the link to the files, and select "Copy Link Address".  
ãƒ•ã‚¡ã‚¤ãƒ«ã¸ã®ãƒªãƒ³ã‚¯ã‚’å³ã‚¯ãƒªãƒƒã‚¯ã—ã€ã€Œãƒªãƒ³ã‚¯ã®URLã‚’ã‚³ãƒ”ãƒ¼ (Copy Link)ã€ã™ã‚‹ã€‚  

### Downloading data
**ãƒ‡ãƒ¼ã‚¿ã®ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰**

ã‚¿ãƒ¼ãƒŸãƒŠãƒ«ã‚’é–‹ãã€‚  
Open a terminal window.  
```
bash

# ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’ä½œæˆã—ã¦ç§»å‹•ã™ã‚‹:
# make and change directory:
mkdir -p ~/projects/data/ensembl
cd ~/projects/data/ensembl/

# `wget` ã‚’ä½¿ç”¨ã—ã¦ã€ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã™ã‚‹:  
# Download the files using `wget`:  
wget https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/CHECKSUMS
wget https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz
wget https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/README

# `zgrep`ã‚³ãƒãƒ³ãƒ‰ã‚’ç”¨ã„ã¦ã€ãƒ‘ã‚¿ãƒ¼ãƒ³ã«ãƒãƒƒãƒã—ãŸè¡Œã‚’æŠ½å‡ºã™ã‚‹:  
# use `zgrep` to extract lines matching the pattern:  
## FASTA header lines begin with the ">" character.
zgrep "^>" Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz

# `sum`ã‚³ãƒãƒ³ãƒ‰ã§ãƒã‚§ãƒƒã‚¯ã‚µãƒ ã‚’è¨ˆç®—ã—ã€CHECKSUMSãƒ•ã‚¡ã‚¤ãƒ«ã®å€¤ã¨æ¯”è¼ƒã™ã‚‹:  
# Calculate checksum using `sum` and compare with values in CHECKSUMS file:
sum Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz
grep "Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz" CHECKSUMS
```

Markdownãƒ•ã‚¡ã‚¤ãƒ« *README.md* ã®ä¾‹:  
The Markdown file *README.md* file would look like:  
```
## Genome Data

Files of mouse (*Mus musculus*) reference genome version GRCm39 (Ensembl release 111) was downloaded on Mon Apr 29 12:58:56 JST 2024, using:

    wget https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/CHECKSUMS
    wget https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz
    wget https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/README

## Checksum

    $sum Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz
    17649 6 Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz

    $grep "Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz" CHECKSUMS
    17649     6 Mus_musculus.GRCm39.dna.chromosome.MT.fa.gz
```

----------
## uniprot_sprot
protein sequence database
ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹

- https://en.wikipedia.org/wiki/UniProt
- https://ja.wikipedia.org/wiki/Swiss-Prot
- 2017.09.12 | 08:03 | UniProtã‚’ä½¿ã£ã¦ã€ã‚¿ãƒ³ãƒ‘ã‚¯è³ªã®ã‚¢ãƒŸãƒé…¸é…åˆ—ã¨ãã®æ©Ÿèƒ½æƒ…å ±ã‚’æ¨ªæ–­çš„ãƒ»ç¶²ç¾…çš„ã«èª¿ã¹ã‚‹ | TogoTV https://doi.org/10.7875/togotv.2017.087

<https://www.uniprot.org/>  

- Open **FTP Download** below **UniProt data**.  
**UniProt data** ã® **FTP Download** ã‚’é–‹ãã€‚
- https://ftp.uniprot.org/pub/databases/uniprot/
  - [knowledgebase/](https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/)
    - [complete/](https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/)
- https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/

      Name                                  Last modified      Size  Description

      RELEASE.metalink                      2024-03-27 10:00  7.9K  

      uniprot_sprot.fasta.gz                2024-03-27 10:00   88M  

- Right click on the link to the files, and select "Copy Link Address".  
ãƒ•ã‚¡ã‚¤ãƒ«ã¸ã®ãƒªãƒ³ã‚¯ã‚’å³ã‚¯ãƒªãƒƒã‚¯ã—ã€ã€Œãƒªãƒ³ã‚¯ã®URLã‚’ã‚³ãƒ”ãƒ¼ (Copy Link)ã€ã™ã‚‹ã€‚  
- https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/reldate.txt
```
UniProt Knowledgebase Release 2024_02 consists of:
UniProtKB/Swiss-Prot Release 2024_02 of 27-Mar-2024
UniProtKB/TrEMBL Release 2024_02 of 27-Mar-2024
```
- https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/  
ãƒ–ãƒ©ã‚¦ã‚¶ï¼ˆFirefox ã¾ãŸã¯ Chromeï¼‰ã§é–‹ãã€‚*uniprot_sprot.fasta.gz* ã‚’å³ã‚¯ãƒªãƒƒã‚¯ã—ã€ã€Œãƒªãƒ³ã‚¯ã®URLã‚’ã‚³ãƒ”ãƒ¼ (Copy Link)ã€ã™ã‚‹ã€‚  
Open the URL with your browser (Firefox or Chrome). Right click the link *uniprot_sprot.fasta.gz* and select "Copy Link Address".  

### Downloading data
**ãƒ‡ãƒ¼ã‚¿ã®ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰**

ã‚¿ãƒ¼ãƒŸãƒŠãƒ«ã‚’é–‹ãã€‚  
Open a terminal window.  
```
# change shell to bash
bash

# ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’ä½œæˆã™ã‚‹
# make directories
mkdir -p ~/projects/data/uniprot/uniprot_sprot

# ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’ç§»å‹•ã™ã‚‹
# change directories
cd ~/projects/data/uniprot/uniprot_sprot/

# *uniprot_sprot.fasta.gz*ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã™ã‚‹
# Download *uniprot_sprot.fasta.gz* file
#curl -O https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/uniprot_sprot.fasta.gz
 wget -b https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/uniprot_sprot.fasta.gz

# `tail -f`ã§ãƒ•ã‚¡ã‚¤ãƒ«å‡ºåŠ›ã‚’ç›£è¦–ã™ã‚‹ï¼ˆControl-Cã§å‹•ä½œä¸­ã®ãƒ—ãƒ­ã‚»ã‚¹ã‚’åœæ­¢ï¼‰
# Use `tail -f` to constantly monitor files (use Control-C to stop)
tail -f wget-log

# "RELEASE.metalink"ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã™ã‚‹
# Download "RELEASE.metalink" file that specifies MD5 checksum https://www.uniprot.org/help/metalink
wget https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/RELEASE.metalink

# MD5ãƒã‚§ãƒƒã‚¯ã‚µãƒ ã‚’è¨ˆç®—ã—ã€"RELEASE.metalink"ãƒ•ã‚¡ã‚¤ãƒ«ã®å€¤ã¨æ¯”è¼ƒã™ã‚‹
# compare our checksum values with those in "RELEASE.metalink" using the md5 program:
md5 uniprot_sprot.fasta.gz
grep -A 3 'file name="uniprot_sprot.fasta.gz"' RELEASE.metalink

# `gunzip`ã‚³ãƒãƒ³ãƒ‰ã§ãƒ•ã‚¡ã‚¤ãƒ«ã‚’å±•é–‹ã™ã‚‹
# decompress files with the command `gunzip`
gunzip -c uniprot_sprot.fasta.gz > uniprot_sprot.fasta
```

Markdownãƒ•ã‚¡ã‚¤ãƒ« *README.md* ã®ä¾‹:  
The Markdown file *README.md* file would look like:  
```
## Files

UniProtKB/Swiss-Prot Release 2024_02 of 27-Mar-2024 was downloaded on 2024-04-29, using:

    wget -b https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/uniprot_sprot.fasta.gz
    wget https://ftp.uniprot.org/pub/databases/uniprot/knowledgebase/complete/RELEASE.metalink

## MD5 checksum

$md5 uniprot_sprot.fasta.gz
MD5 (uniprot_sprot.fasta.gz) = 0766df3e5785fc5f1cfc496aa89e86ad

$grep -A 3 'file name="uniprot_sprot.fasta.gz"' RELEASE.metalink
  <file name="uniprot_sprot.fasta.gz">
      <size>92181645</size>
      <verification>
        <hash type="md5">0766df3e5785fc5f1cfc496aa89e86ad</hash>
```

### Inspecting data
ãƒ‡ãƒ¼ã‚¿ã®æ¤œæŸ»

Using core Unix tools (`head, tail, wc, ls`) to analyze the sequence data. 
Unixã‚³ãƒãƒ³ãƒ‰ï¼ˆ`head, tail, wc, ls`ï¼‰ã‚’ç”¨ã„ã¦ã€é…åˆ—ãƒ‡ãƒ¼ã‚¿ã‚’è§£æã™ã‚‹ã€‚  

```
# `ls -lh`ã§ãƒ•ã‚¡ã‚¤ãƒ«ã®ã‚µã‚¤ã‚ºã‚’ç¢ºèªã™ã‚‹:  
# `ls -lh` reports human-readable file sizes
ls -lh

# å¤‰æ•°ã«å€¤ã‚’å‰²ã‚Šå½“ã¦ã‚‹ï¼ˆ`=`ã®å‰å¾Œã«ã‚¹ãƒšãƒ¼ã‚¹ã‚’å…¥ã‚Œãªã„ï¼‰:  
# create a variable and assign it a value with (do not use spaces around the equals sign!):  
MYSEQ=uniprot_sprot.fasta

# å¤‰æ•°ã®å€¤ã«ã‚¢ã‚¯ã‚»ã‚¹ã™ã‚‹ã«ã¯ã€å¤‰æ•°åã®å‰ã«ãƒ‰ãƒ«è¨˜å·ã‚’ä»˜ã‘ã‚‹:  
# To access a variableâ€™s value, we use a dollar sign in front of the variableâ€™s name:  
echo "${MYSEQ}"

# `head`ã§å…ˆé ­éƒ¨åˆ†ã‚’è¡¨ç¤ºã™ã‚‹:  
# look at the top of a file
head -n 2 "${MYSEQ}"

# `tail`ã§æœ«å°¾éƒ¨åˆ†ã‚’è¡¨ç¤ºã™ã‚‹:  
# look at the end of a file
tail -n 3 "${MYSEQ}"

# `grep`ã§ãƒ‘ã‚¿ãƒ¼ãƒ³ã«ãƒãƒƒãƒã™ã‚‹è¡Œã‚’æŠ½å‡ºã™ã‚‹ï¼ˆControl-Cã§å‹•ä½œä¸­ã®ãƒ—ãƒ­ã‚»ã‚¹ã‚’åœæ­¢ï¼‰
# use `grep` to extract lines matching the pattern "^>" (use Control-C to stop)
## FASTA header lines begin with the ">" character.
grep "^>" "${MYSEQ}"

# ãƒ‘ã‚¤ãƒ—ã§ãƒ—ãƒ­ã‚°ãƒ©ãƒ ã®å…¥å‡ºåŠ›ã‚’ã¤ãªã
# Pipe the standard output to the next command with the pipe character (`|`).
grep "^>" "${MYSEQ}" | head

# `wc -l`ã§è¡Œæ•°ã‚’ã‚«ã‚¦ãƒ³ãƒˆã™ã‚‹
# `wc -l` outputs the number of lines
grep "^>" "${MYSEQ}" | wc -l

# `grep -c`ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã§ã€ãƒ‘ã‚¿ãƒ¼ãƒ³ã«ãƒãƒƒãƒã—ãŸè¡Œæ•°ã‚’è¡¨ç¤ºã™ã‚‹:  
# use grep to count (the -c option stands for count) the number of lines matching the pattern:  
grep -c "^>" "${MYSEQ}"

# `grep -i`ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã§ã€å¤§æ–‡å­—å°æ–‡å­—ã‚’åŒºåˆ¥ã—ãªã„ï¼ˆignore caseï¼‰
# add the option `-i` to `grep` to be case insensitive.
grep "^>" "${MYSEQ}" | grep -i "Shugoshin" | wc -l

# Parsing FASTA headers using perl
NUM=10
grep '^>' "${MYSEQ}" | perl -ne '$_=~/^>(\S+) (.+) OS=(.+?) (GN|PE)=/; print "$1 ~~ $2 ~~ $3\n";' | head -n "${NUM}"
echo; echo "# ${NUM} most abundant OrganismName"
grep '^>' "${MYSEQ}" | perl -nle '$_=~/^>(\S+) (.+) OS=(.+?) (GN|PE)=/; print "$3";' | sort | uniq -c | sort -nr | head -n "${NUM}"
echo; echo "# ${NUM} most abundant ProteinName"
grep '^>' "${MYSEQ}" | perl -nle '$_=~/^>(\S+) (.+) OS=(.+?) (GN|PE)=/; print "$2";' | sort | uniq -c | sort -nr | head -n "${NUM}"
```

----------

https://github.com/haruosuz/bioinfo/blob/master/2019/CaseStudy.md

## NCBI ASSEMBLY_REPORTS

[NCBI](https://ja.wikipedia.org/wiki/å›½ç«‹ç”Ÿç‰©å·¥å­¦æƒ…å ±ã‚»ãƒ³ã‚¿ãƒ¼)ã®ã‚²ãƒãƒ é…åˆ—ã®ãƒ¡ã‚¿ãƒ‡ãƒ¼ã‚¿ãŒè¨˜è¼‰ã•ã‚Œã¦ã„ã‚‹ *assembly_summary* ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ç”¨ã„ã¦ã€ç›®çš„ã®ã‚²ãƒãƒ é…åˆ—ã‚’è¦‹ã¤ã‘ã‚‹ã€‚  

- [NCBI Genomes FTP](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/policies-annotation/genomeftp/)
- [Metagenomics - NCBI ftp genome download](https://www.metagenomics.wiki/tools/fastq/ncbi-ftp-genome-download)
How to download all reference genomes of a selected species from NCBI    

<https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/>

URL ã‚’ãƒ–ãƒ©ã‚¦ã‚¶ï¼ˆFirefox ã¾ãŸã¯ Chromeï¼‰ã§é–‹ãã€‚*README_assembly_summary.txt*ã‚’ã‚¯ãƒªãƒƒã‚¯ã™ã‚‹ã€‚  
Open the URL with your browser (Firefox or Chrome). Click the link *README_assembly_summary.txt*.  

```
# https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/README_assembly_summary.txt

The assembly_summary files report metadata for the genome assemblies on the NCBI genomes FTP site.

assembly_summary_genbank.txt            - current GenBank genome assemblies

assembly_summary_refseq.txt             - current RefSeq genome assemblies

https://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank.txt
https://ftp.ncbi.nlm.nih.gov/genomes/refseq/assembly_summary_refseq.txt
```

- January 31, 2020. [What is the difference between RefSeq and GenBank?](https://www.ncbi.nlm.nih.gov/books/NBK50679/#RefSeqFAQ.what_is_the_difference_between_1)
- 2023-08-31 [RefSeq - JI](https://fish-evol.org/RefSeq.html)
- 2020.04.18 [RefSeq | è©³ç´°ãªæ³¨é‡ˆã¥ã‘ã‚‰ã‚Œã¦ã„ã‚‹å†—é•·æ€§ã®ãªã„æ ¸é…¸ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹](https://bi.biopapyrus.jp/db/refseq.html)

----------
























